doc/design/bucketpath.rst

```diff
@@ -0,0 +1,157 @@
+==============================
+Design Document Bucket Path(s)
+==============================
+
+Introduction
+============
+
+- Purpose of the document / Problem description
+- Definitions and Acronyms
+
+Overview
+========
+
+Design Goals
+============
+
+Path Design
+===========
+
+Future Considerations
+=====================
+
+Sudo Code Outline
+=================
+
+.. code-block:: python
+
+    from exasol.bucketfs import PathBuilder
+
+    Path = PathBuilder(credentials)
+
+    udf_path = Path("bfsl://some/local/path/file.tar.gz")
+    http_bucket_path = Path("bfs://127.0.0.1:8888/service/bucket/some/file.tar.gz")
+    https_bucket_path = Path("bfss://127.0.0.1:8888/service/bucket/some/file.tar.gz")
+    chroot_path = Path("bfss://127.0.0.1:8888/service/bucket/some/sub/subsub/file.tar.gz?chroot=/some/sub/")
+
+
+
+.. code-block:: python
+
+    from typing import Protocol
+
+    # interface / protocol or abc
+    # Note: should mimic pathlib.Path so it can easily understood and adopted. 
+    class Pathlike(Protocol):
+
+        @property
+        def name:
+            ...
+
+        @property
+        def suffix:
+            ...
+
+        @property
+        def root:
+            ...
+
+        def as_uri():
+            ...
+
+        def exists():
+            ...
+
+        def is_dir():
+            ...
+
+        def is_file():
+            ...
+
+        def read_text():
+            ...
+
+        def read_bytes():
+            ...
+
+        def write_text():
+            ...
+
+        def write_bytes():
+            ...
+
+        def unlink():
+            ...
+
+        def rmdir():
+            ...
+
+        def join()
+            ...
+
+        def walk():
+            ...
+
+        def iterdir():
+            ...
+
+        # overload / for joining
+        def __truediv__():
+            ...
+
+
+    # Note: needs to support Pathlike
+    class BucketPath:
+        
+        protocol = ['bfs', 'bfss']
+
+        def __init__(bucket: Bucket, path):
+            ...
+
+
+    # QUESTION: UdfPath? Do we need to distuinguish between localpaths and localpaths within udf?
```

Comment:
UDFPath can't write, besides that both operate on local file system. Only the local path we create might be different. For testing it points to a temp dir and for udfs in points to `/buckets/<service>/<bucket>/...`

Reply:
TL;DR: Local paths are working exactly the same except they having a different root? (So the Chroot may still come in handy here)

Reply:
Actually a good point. We could setup the same structure as in the UDF in a temporary directory. For example:

`/tmp/tmpdir/buckets/default/path/in/bucket`